run.c

```diff
@@ -315,6 +315,11 @@ void compute_accelerations(int is_PM, int FirstStep)
 
     walltime_measure("/Misc");
 
+    grav_short_tree();
+
+    /* PM is after tree as we need GravPM
+     * from the last step to compute the opening
+     * angle in tree.*/
```

Comment:
This breaks the formal agreement between the KkddkK and the code -- making it kKddkK.

Since the opening condition introduces a higher order error, we probably can either use the old PM or neglect PM force alltogether in the condition -- can we save both to the snapshot and look at the statistics?

Reply:
We currently use old PM: the PM over-writes GravPM, so to use the old PM in tree we need the PM step to come after tree. Note that no drifting or kicking is done in the PM or tree code anymore: we are just computing the accelerations and potential. So the formal properties are unchanged, but obscured.

I thought about not using PM, or using new PM, but while the timestep is set so that the acceleration does not change too much, it does not make the same guarantee about GravAccel and GravPM: consider two particles separated by ~Asmth: the acceleration will be roughly the same between timesteps but it may shift quickly from GravAccel to GravPM, meaning that using one of them from a different timestep may introduce a large error.

If you are unhappy with the ordering changing, we should just close this pull request and forego the optimization.

Reply:
Does the change of ordering actually work? PM frees the tree (reusing the memory for mesh), but both grav and hydro needs the tree.  So after the ordering change we are building the tree twice?

On your example:
You forgot the long/short range smoothing kernel and Acut. If two particles separated by ~Asmth, then the force will be dominated by tree because Asmth << Acut. If two particles are separated by ~Acut, then the force will be dominated by PM because the short range at Acut is suppressed by erf.

Reply:
You are right there is a sweet spot where the force is half PM and half Tree, but it shouldn't suddenly jump between the two because of the smoothing. 

I tend to view short range and long range gravity as two separate forces (because the code says so), thus the accuracy of short range shall only depend on short range force. But the argument can also be made that we want to control the accuracy of the total gravity, because the physics says so. However the physics also says the total force includes hydro and effective forces from the feedback, and we don't take care of those here neither.

Reply:
I think we rebuild the tree twice on a PM step anyway, don't we? We rebuild it at the end of the domain decomposition, use it to do PM assignment to regions, free it, and then rebuild it again.
Moving the PM later would actually allow us to only rebuild it once: we could in principle do hydro and feedback, then do PM *at the end* and avoid rebuilding it after we free. But this is probably too much of a contortion for a small gain: the tree rebuild is a quite small fraction of a PM step.

Reply:
The code is allowed to make the timestep long enough that the particle can move across the Acut/Asmth boundary, especially if the particle is moving fast and has small acceleration. It is because the timestep is set by the total force.

I agree that arguably either the short timestep should just be GravAccel, or both timestep and truncation should include hydro. Actually that seems a good short paper for a beginning student...

Reply:
You are right. But moving all shortrange before PM looks cleaner than having some of them before and some after. Could you add a few words in the comment to say the purpose of using old PM acc is to maintain agreement with Gadget2?

Reply:
Absolutely.

Reply:
Let's make sure the options are documented in the code such that the next person reading it knows why it is this way but not the other way.